### ISSUE TITLE
Document Backup and Restore

### ISSUE THREAD

**Issue by rvanlaar** [2ndQuadrant/bdr#140]
Currently there is no documentation about how to backup and restore a bdr database.

Specifically, I would like documentation on:
how to generate a dump with bdr_dump
restoring a bdr database to a non bdr database
restoring a bdr database to a new bdr cluster
how to restore sequences

**Comment by sknohsalla** [comment]
+1

**Comment by dwierenga** [comment]
+1

Even if these things aren't generally used as much for multi-master setups, they go a long way for the comfort level and confidence of the DBA.

**Comment by lbdremy** [comment]
Indeed that's a must have, I really looking forward.
In my experience, I had no trouble to dump a simple postgres database and restore it into a bdr database but I'm having trouble to dump a bdr database and restore it into an existing bdr cluster with only the database created.

**Comment by rvlander** [comment]
Same thing here.

**Comment by rvanlaar** [comment]
I followed the following course of action for creating restorable dumps.
My main pain points were the inclusion of bdr related tables in my dumps
and sequences that were unrestorable. I fixed them the following way.

My settings include default_sequenceam = 'bdr' for the sequences.

## create a schema only and data only dump

```
/usr/lib/postgresql/9.4/bin/bdr_dump -Fp -h localhost -U postgres DDNAME -f /tmp/data.sql --data-only
/usr/lib/postgresql/9.4/bin/bdr_dump -Fp -h localhost -U postgres DBNAME -f /tmp/schema.sql --schema-only
```

## remove sequence Creation from schema

The sed command explained:
1. Remove all create sequence parts
2. Remove all alter commands for sequences
3. Remove all triggers, they are created automatically

```
sed -e '/CREATE SEQUENCE/ { N;N;N;N;N;N; d; }' < /tmp/schema.sql  |grep -v '_id_seq' |grep -v 'truncate_trigger' > /tmp/nobdr-schema.sql
```

## remove bdr data from the data dump

The sed command explained:
1. Remove all "COPY bdr_" parts from the data dump
```
sed '/^COPY bdr_/,/^\\\.$/d' /tmp/data.sql  > /tmp/nobdr-data.sql
```

## Generate sql code to start the sequence at the correct value: START WITH VALUE
Note: All my sequences end with _id_seq.
1. Get all the sequences from the schema
2. Use sed and awk to print the correct sql
3. output it to a new file
```
grep 'CREATE SEQUENCE' /tmp/schema.sql |awk '{print $3}'|sed 's/_id_seq$//' | awk '{print "SELECT '\''CREATE SEQUENCE " $1"_id_seq START WITH '\'' || COALESCE(MAX(id)+1,1) || '\'';'\'' FROM " $1 ";"}'  > /tmp/get-sequences.sql
```
## execute the generated SQL and append the output to nobdr-schema.sql
```
psql -U postgres DBNAME -Atq -f /tmp/get-sequences.sql >> /tmp/nobdr-schema.sql
```

## clean-up the get-sequences file
```
rm /tmp/get-sequences.sql
```

## Append all sequence settings (table, owner) to the new schema file.
grep '_id_seq' /tmp/schema.sql |grep -v '\--' |grep -v CREATE >> /tmp/nobdr-schema.sql

## Voila, restore your data this way on a new bdr cluster or a regular postgres database
```
psql -h localhost -U postgres -d DBNAME -f nobdr-schema.sql
psql -h localhost -U postgres -d DBNAME -f nobdr-data.sql
```

**Comment by lbdremy** [comment]
So nice of you @rvanlaar :=)

**Comment by lbdremy** [comment]
Thanks again @rvanlaar , I'm running into an error:

```
psql:/tmp/bdr_backup/backup.sql:3449: ERROR:  cannot call setval() on global sequence public.user_id_seq
```
And I have this error for some of my sequences until I get : 

```
psql:/tmp/bdr_backup/backup.sql:2402395: WARNING:  terminating connection because of crash of another server process
DETAIL:  The postmaster has commanded this server process to roll back the current transaction and exit, because another server process exited abnormally and possibly corrupted shared memory.
HINT:  In a moment you should be able to reconnect to the database and repeat your command.
```

In which circustances to you run these commands ? For my part I'm running this on a bdr cluster with 3 nodes all ready.

I have made this script from your comment: https://gist.github.com/lbdremy/78b119116c8daf20528a

**Comment by lbdremy** [comment]
And in order to provision we use this command:
```
ssh $SSH_USER@$HOST "sudo -u postgres psql -d $DATABASE_NAME -f /tmp/$BACKUP_NAME/backup.sql"
```

**Comment by rvanlaar** [comment]
@lbdremy setval is not permitted on BDR, that's why the START WITH VALUE is used. 

This [line](https://gist.github.com/lbdremy/78b119116c8daf20528a#file-backup-bdr-sh-L32)
removes all occurrences of _id_seq out the original schema dump.

Just execute that one and check if there still is a set value in you dump.

**Comment by lbdremy** [comment]
Yes @rvanlaar , but some statements `SELECT pg_catalog.setval()` were added back on this line  https://gist.github.com/lbdremy/78b119116c8daf20528a#file-backup-bdr-sh-L41 so remove them on this line. I'm still working on it, I run into another issue no disk space left creating a bigger VM now... :) Thanks @rvanlaar

**Comment by ringerc** [comment]
While the docs very badly need improving here, there are also improvements needed in the underlying code. Some of which really require core patches, and some of of which aren't clear-cut at all.

Docs need to cover:

- How to restore from a physical backup. Strip out BDR, re-create as new BDR cluster or extract the database to restore to a standalone stock Pg.

- How to restore from a logical backup (pg_dump) to either a new BDR cluster on a patched Pg or to a standalone DB on a stock Pg.

**Comment by qnub** [comment]
@rvanlaar thank you very much! I just have not enough words to say how grateful i'm to you! :clap:

**Comment by dark-light-cz** [comment]
@rvanlaar thank you very much! Great work! This is only one usable information on restoring BDR database. 👍

**Comment by markjcrane** [comment]
How to strip out bdr use pg_dump to backup the public schema in other words leaving out the bdr schema.

**Comment by dwierenga** [comment]
@markjcrane I simply use `bdr_dump -Fp -npublic $DBNAME | perl -p -i -e "s/    USING bdr;/    USING local;/" | gzip -9 > ${DBNAME}.dump.sql.gz`
This leaves me with a SQL file that can import correctly to a database that doesn't have the BDR extension installed - though you still need the BDR-patched binaries.  If you wanted to import to a stock Postgres install, then changing the perl command to `perl -p -i -e "s/    USING bdr;/;/"` should do it, though I haven't tested that.

**Comment by rvanlaar** [comment]
@markjcrane The backup script removes all BDR references. Just removing 'USING bdr' doesn't set the sequences.

The @dwierenga does `USING local` work with the sequences?

**Comment by dwierenga** [comment]
@rvanlaar : Yes, it actually works.  The dump command only dumps the public schema, so I avoid all of the BDR objects and the `create extension` statements.  The actual sequence creation statements contains a `START WITH` clause so they have the initial value one would expect.

**Comment by rvanlaar** [comment]
@dwierenga Does it also work restoring it to a bdr database?

**Comment by dwierenga** [comment]
@rvanlaar IIRC it had issues restoring if BDR was enabled prior to running the restore.  I don't remember exactly what they were, but I didn't investigate too much because I figured if I already had BDR up and running, I'm not restoring... 

My use case was generally bootstrapping a new cluster - either a copy of production to a dev environment, or nuking a broken BDR configuration and reloading my production data.

**Comment by rvanlaar** [comment]
@dwierenga One thing the backup script at the top does is getting the current values for sequences and setting it explicitly. 

Restoring a BDR database to a BDR cluster doesn't allow for a sequence setval.

**Comment by ringerc** [comment]
I'm working on improving the situation with dump-based backup and restore, either for 1.0 or (more likely at this point) just after 1.0.

There are really two target use cases:

* Take a BDR DB, restore to stock PostgreSQL as a standalone DB. Here we don't want the BDR schema and extension, don't want USING syntax in sequence creation, want to omit the BDR security labels, etc. Global sequences should be converted to local sequences, but with the start value set to the highest allocated global value on any node ( related to #215 ).

* Take a BDR DB and use it to create a new BDR group. In this case we want to restore global sequences as global sequences, retaining their allocations. We want to retain the security labels on the tables (but see #225, we don't want the database label ), the bdr schema's global sequence state, etc. But we want to omit bdr_nodes, bdr_connections, etc. (We can't just exclude these from dumps unconditionally, since we currently require them for logical init).

Note that you *cannot* restore a dump of a BDR node to rejoin a failed node to an existing BDR group. The replication identifiers and slots will be missing.

A lot of the underlying mess is created by BDR's use of `pg_dump` and `pg_restore` as part of its node bringup process, because PostgreSQL its self lacks the kind of schema-copy and introspection features we'd need to avoid doing so.

**Comment by smaslennikov** [comment]
@ringerc Has there been any progress on this? My case is the first one: dumping a BDR instance and importing it into CloudSQL. Here are things I've tried:

- `bdr_dump` doesn't have [some needed flags for the operation](https://cloud.google.com/sql/docs/postgres/import-export/exporting): `--format=plain --no-owner --no-acl`, though @rvanlaar 's method seemed to set us in the right direction;
- using a vanilla (unpatched) `pg_dump` (as described in https://github.com/2ndQuadrant/bdr/issues/201) on another system (jumpbox's `pg_dump` hitting a pg server, piping into psql to CloudSQL) doesn't work either - bdr data is still seemingly exported

What other methods are there to try?

**Comment by smaslennikov** [comment]
@ringerc is there a sensible way of doing that, aside from either dumping table by table or editing the dump later? Some of our affected databases are going to be quite large.

What exactly needs to be removed - any lines, tables, extensions mentioning BDR?

**Comment by ringerc** [comment]
On 20 July 2018 at 01:30, Slava Maslennikov <notifications@github.com>
wrote:

> @ringerc <https://github.com/ringerc> is there a sensible way of doing
> that, aside from either dumping table by table or editing the dump later?
> Some of our affected databases are going to be quite large.
>

Use pg_restore's Table of Contents filtering features.

See pg_restore --list and pg_restore --use-list

It's clumsy and not ideal. I'd really welcome a patch to add an
extension-exclude mode to pg_dump.

-- 
 Craig Ringer                   http://www.2ndQuadrant.com/
 PostgreSQL Development, 24x7 Support, Training & Services

**Comment by ringerc** [comment]
On 20 July 2018 at 09:11, Craig Ringer <craig@2ndquadrant.com> wrote:

> On 20 July 2018 at 01:30, Slava Maslennikov <notifications@github.com>
> wrote:
>
>> @ringerc <https://github.com/ringerc> is there a sensible way of doing
>> that, aside from either dumping table by table or editing the dump later?
>> Some of our affected databases are going to be quite large.
>>
>
> Use pg_restore's Table of Contents filtering features.
>
> See pg_restore --list and pg_restore --use-list
>
> It's clumsy and not ideal. I'd really welcome a patch to add an
> extension-exclude mode to pg_dump.
>


Alternately, you can limit pg_dump to your application schemas.


-- 
 Craig Ringer                   http://www.2ndQuadrant.com/
 PostgreSQL Development, 24x7 Support, Training & Services

**Comment by smaslennikov** [comment]
@ringerc Thank you for your help. I ended up putting a [bash script](https://github.com/smaslennikov/smaslennikov.github.io/blob/master/bin/migrate_bdr_to_postgres.sh) together for this purpose. I'll be perfecting it during several migrations in the next week.
